File: leetcode/201-300/283-move-zeroes/bench/movezero.c
/* LeetCode 283 bench mirror — C. Same write cursor, same per-round refresh,
 * same position-weighted sink plus write count. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#define N 2000000
#define ROUNDS 60
static void move_zeroes(int64_t *a, int64_t n, int64_t *stores) {
    int64_t write = 0;
    for (int64_t i = 0; i < n; i++)
        if (a[i] != 0) { a[write] = a[i]; (*stores)++; write++; }
    while (write < n) { a[write] = 0; (*stores)++; write++; }
}
int main(void) {
    int64_t *src = malloc((size_t)N*sizeof(int64_t));
    int64_t *work = malloc((size_t)N*sizeof(int64_t));
    int64_t seed = 20260821;
    for (int64_t i = 0; i < N; i++) {
        seed = (seed*1103515245LL+12345LL)%2147483648LL;
        src[i] = (seed % 2 == 0) ? 0 : seed % 100003LL;
    }
    int64_t sink = 0, total = 0;
    for (int r = 0; r < ROUNDS; r++) {
        memcpy(work, src, (size_t)N*sizeof(int64_t));
        int64_t st = 0;
        move_zeroes(work, N, &st);
        total += st;
        int64_t h = 0;
        for (int64_t j = 0; j < N; j++) h = (h*31 + work[j]) % 1000000007LL;
        sink = (sink + h) % 1000000007LL;
    }
    printf("%lld %lld\n", (long long)sink, (long long)total);
    return 0;
}

File: leetcode/201-300/283-move-zeroes/bench/movezero.rs
// LeetCode 283 bench mirror — Rust. Same cursor, refresh and sink.
const N: usize = 2000000;
const ROUNDS: usize = 60;
fn move_zeroes(a: &mut [i64], stores: &mut i64) {
    let mut write = 0usize;
    for i in 0..a.len() {
        if a[i] != 0 { a[write] = a[i]; *stores += 1; write += 1; }
    }
    while write < a.len() { a[write] = 0; *stores += 1; write += 1; }
}
fn main() {
    let mut seed: i64 = 20260821;
    let mut src = vec![0i64; N];
    for i in 0..N {
        seed = (seed*1103515245+12345)%2147483648;
        src[i] = if seed % 2 == 0 { 0 } else { seed % 100003 };
    }
    let mut work = vec![0i64; N];
    let (mut sink, mut total) = (0i64, 0i64);
    for _ in 0..ROUNDS {
        work.copy_from_slice(&src);
        let mut st = 0i64;
        move_zeroes(&mut work, &mut st);
        total += st;
        let mut h = 0i64;
        for j in 0..N { h = (h*31 + work[j]) % 1000000007; }
        sink = (sink + h) % 1000000007;
    }
    println!("{} {}", sink, total);
}
